tracer/src/Datadog.Trace/Agent/StatsAggregator.cs

```diff
@@ -159,15 +167,17 @@ private static double ConvertTimestamp(long ns)
             return ns << shift;
         }
 
-        private void AddToBuffer(Span span)
+        private bool AddToBuffer(Span span)
         {
             if ((!span.IsTopLevel && span.GetMetric(Tags.Measured) != 1.0) || span.GetMetric(Tags.PartialSnapshot) > 0)
             {
-                return;
+                return false;
             }
```

Comment:
[Looking at the RareSampler](https://github.com/DataDog/datadog-agent/blob/f0b45d1b84937f7ab43078b25f27be7124b3eb15/pkg/trace/sampler/rare_sampler.go#L103), the sampler only runs on top-level/measured spans, so we're consistent there.

As for the error part, I think you've identified a bug in the Java implementation. That's the reference implementation I followed, where errors are only checked on top-level/measured spans. However, the agent [checks for errors on _any_ span](https://github.com/DataDog/datadog-agent/blob/9f6da352188ce778d743f8c1c5e07cc7b630dd72/pkg/trace/agent/agent.go#L486-L488) in the trace chunk. I will revise this.

Reply:
Thinking on this one particular issue more, the bug you identified was that we could return `false` even when there's an error. However, I don't think this is actually an issue. This return value is only used as a fallback after the `TraceContext` ran its rules. When the `TraceContext` closed the trace and sent it to serialization, we would have already detected that there was at least one error within the trace so we would have called `AgentWriter.SerializeTrace(ArraySegment<Span> trace, bool shouldSerializeSpans)` with the value `true` which would disregard the value returned here and send the trace to the trace agent. So, this is only a fallback and the initial decision resulted in sending the error-containing trace.

This will go away soon with the new refactor, but I thought it was worth reporting